File: python/find_lambda.py
#!/usr/bin/env python3
import sys
import math
import numpy as np

def epsilon(d, theta1, theta2, lam):
    return math.log((d + (1 + theta1) * lam) / ((1 - theta2) * lam))

def delta(d, theta1, theta2, lam):
    term1 = math.exp(- (theta1 * theta1 * lam) / ((2 + theta1) * d))
    term2 = math.exp(- (theta2 * theta2 * lam) / (2 * d))
    return term1 + term2

def find_min_lambda(d, eps_max, delta_max, theta1_max, 
                    theta_step=0.01, lambda_max=10**9):
    theta_values = np.arange(0.0, theta1_max + 1e-12, theta_step)
    theta_values_one = np.concatenate([
        np.arange(0.0, 1.0 - 1e-12, theta_step),
        np.array([1 - 10**(-k) for k in range(3, 10)])
    ])

    best_lambda = None
    best_tuple = None

    for theta1 in theta_values:
        for theta2 in theta_values_one:

            # Exponential search for lambda
            lam = 1
            ok = False
            while lam <= lambda_max:
                eps = epsilon(d, theta1, theta2, lam)
                delt = delta(d, theta1, theta2, lam)

                if eps <= eps_max and delt <= delta_max:
                    ok = True
                    break
                lam *= 2  # accelerate search

            if not ok:
                continue

            # Binary search between lam/2 and lam
            low = lam // 2
            high = lam
            while low + 1 < high:
                mid = (low + high) // 2
                if (epsilon(d, theta1, theta2, mid) <= eps_max and
                    delta(d, theta1, theta2, mid) <= delta_max):
                    high = mid
                else:
                    low = mid

            lam_candidate = high

            if best_lambda is None or lam_candidate < best_lambda:
                best_lambda = lam_candidate
                best_tuple = (theta1, theta2)

    return best_lambda, best_tuple


def main():
    # ---- parse argv ----
    if len(sys.argv) < 4:
        print("Usage: python find_lambda.py <d> <eps_max> <delta_max> (<theta1-max (default: 2.0)>)")
        sys.exit(1)

    d = int(sys.argv[1])
    eps_max = float(sys.argv[2])
    delta_max = float(sys.argv[3])
    
    theta1_max = 2.0
    if len(sys.argv) >= 5:
        theta1_max = float(sys.argv[4])

    # ---- compute ----
    lam, (theta1, theta2) = find_min_lambda(d, eps_max, delta_max, theta1_max)

    # ---- output ----
    print("lambda =", lam)
    print("best theta1 =", theta1)
    print("best theta2 =", theta2)


if __name__ == "__main__":
    main()

File: java/fodp/src/fodp/FindLambda.java
package fodp;

public class FindLambda {

	public static double epsilon(int d, double xi1, double xi2, long lam) {
		return Math.log((d + (1 + xi1) * lam) / ((1 - xi2) * lam));
	}

	public static double delta(int d, double xi1, double xi2, long lam) {
		double term1 = Math.exp(-(xi1 * xi1 * lam) / ((2 + xi1) * d));
		double term2 = Math.exp(-(xi2 * xi2 * lam) / (2 * d));
		return term1 + term2;
	}

	public static class Result {
		long lambda;
		double xi1;
		double xi2;

		public Result(long lambda, double xi1, double xi2) {
			this.lambda = lambda;
			this.xi1 = xi1;
			this.xi2 = xi2;
		}
	}

	public static long findMinLambda(int d, double epsMax, double deltaMax) {
		return findMinLambda(d, epsMax, deltaMax, 0.001, 1000000000000L).lambda;

	}

	public static Result findMinLambda(int d, double epsMax, double deltaMax, double xiStep, long lambdaMax) {
		Long bestLambda = null;
		Double bestXi1 = null;
		Double bestXi2 = null;

		// Generate xi values
		for (double xi1 = xiStep; xi1 <= 1.0 + 1e-12; xi1 += xiStep) {
			for (double xi2 = xiStep; xi2 <= 1.0 - 1e-12; xi2 += xiStep) {
				// Exponential search for λ
				long lam = 1;
				boolean ok = false;

				while (lam <= lambdaMax) {
					double eps = epsilon(d, xi1, xi2, lam);
					double delt = delta(d, xi1, xi2, lam);

					if (eps <= epsMax && delt <= deltaMax) {
						ok = true;
						break;
					}
					lam *= 2; // accelerate search

					// Prevent overflow
					if (lam < 0) {
						lam = lambdaMax;
						break;
					}
				}

				if (!ok) {
					continue;
				}

				// Binary search between lam/2 and lam
				long low = lam / 2;
				long high = lam;

				while (low + 1 < high) {
					long mid = (low + high) / 2;
					if (epsilon(d, xi1, xi2, mid) <= epsMax && delta(d, xi1, xi2, mid) <= deltaMax) {
						high = mid;
					} else {
						low = mid;
					}
				}

				long lamCandidate = high;

				if (bestLambda == null || lamCandidate < bestLambda) {
					bestLambda = lamCandidate;
					bestXi1 = xi1;
					bestXi2 = xi2;
				}
			}
		}

		return new Result(bestLambda, bestXi1, bestXi2);
	}

	public static void main(String[] args) {

		int d = 18201;
		double epsMax = 10;
		double deltaMax = 1E-13;

		// Compute
		Result result = findMinLambda(d, epsMax, deltaMax, 0.001, 1000000000000L);

		// Output
		System.out.println("lambda = " + result.lambda);
		System.out.println("best xi1 = " + result.xi1);
		System.out.println("best xi2 = " + result.xi2);

		double epsilon = Math.log((d + (1 + result.xi1) * result.lambda) / ((1 - result.xi2) * result.lambda));
		System.out.println(epsilon);
	}
}